### ISSUE TITLE
[Feature Request] Make it easier to override snippet admin views

### ISSUE THREAD

**Issue by juan0tron** [wagtail/wagtail#5674]
### Is your proposal related to a problem?

One of my biggest time wasters in new wagtail projects is setting up custom templates for snippet admin views. By default, `list.html` only shows the `__str__` attribute of each snippet. I often find customizing this template to be very valuable to my users, but overriding them requires copying a significant amount of Wagtail's source code.

Honestly, I'm not sure if my solution to this is just plain the wrong way to go about it, but I haven't seen much documentation describing an easier way to do it.

### Describe the solution you'd like

This could be solved one of two ways:

- Provide a `columns` attribute in snippet models, similar to `panels`, that would allow users to define what fields are displayed in the `list.html` template

```
@register_snippet
class Foo(index.Indexed, ClusterableModel):

    bar = models.CharField(max_length=255)

    columns = [
        'bar',
        'a_custom_method'
    ]

    panels = [
        FieldPanel('bar')
    ]

    def a_custom_method(self):
        # some logic
        return 'some custom string defined by this method's logic'
```

Then, `list.html` could be updated to loop through these fields and add additional <th> and <td> columns to the listing view based on these fields.

- Provide a `admin_templates` object that allows users to define paths to custom templates for their snippet model:

```
admin_templates = [
    ('list', 'foo/list.html'),
    ('type_index', foo/type_index.html)
]
```

This could be useful in scenarios where the developer wants to add specialized markup to their templates (such as thumbnails) and not just the contents of certain fields in their model.

### Describe alternatives you've considered

- I've tried simply overriding the default templates by placing them in my project under `<APP_NAME>/wagtailsnippets/snippets/<TEMPLATE_NAME>`, but this only works if I want to override that template for ALL of my snippets. This means I'd have to stuff a specific override of the template with if/else statements depending on the model name, which isn't very elegant.
- What DOES seem to work for me is creating custom versions of many of the functions in `snippets.py` and overriding the logic that specifies what template those views use. However, this requires I also create a custom URL to override the default URL for that snippet. Until recently I had a custom `list()` view and URL for every single snippet in my project, which wasn't very DRY.

### Additional context

Here's an example of my custom snippet `list()` function (initially copied from wagtail source and customized by my team over the past few months). This is in a custom `snippet` app I use in my project that allows other templates to extend the default snippet views without repeating much code. In this case I've set this function up to:

- Look for templates under `snippet_name/<TEMPLATE_NAME>.html` (instead of  `wagtailsnippets/snippets/<TEMPLATE_NAME>.html`)
- Include an `add_context` argument to allow my snippets to pass their own context variables to snippet views in their respective `list()` functions.

```
def list_snippets(request, snippet_type, snippet_model, add_context={}):
	model       = get_snippet_model_from_url_params(snippet_type, snippet_model)
	permissions = [
		get_permission_name(action, model)
		for action in ['add', 'change', 'delete']
	]
	if not any([request.user.has_perm(perm) for perm in permissions]):
		return permission_denied(request)

	items = model.objects.all()

	# Preserve the snippet's model-level ordering if specified, but fall back on PK if not
	# (to ensure pagination is consistent)
	if not items.ordered:
		items = items.order_by('pk')

	# Search
	is_searchable = class_is_indexed(model)
	is_searching = False
	search_query = None
	if is_searchable and 'q' in request.GET:
		search_form = SearchForm(request.GET, placeholder=_("Search %(snippet_type_name)s") % {
			'snippet_type_name': model._meta.verbose_name_plural
		})

		if search_form.is_valid():
			search_query = search_form.cleaned_data['q']

			search_backend = get_search_backend()
			items = search_backend.search(search_query, items)
			is_searching = True

	else:
		search_form = SearchForm(placeholder=_("Search %(snippet_type_name)s") % {
			'snippet_type_name': model._meta.verbose_name_plural
		})

	# Determine user's edit permissions for each item so we can grey out the ones that are inaccessible
	for item in items:
		if(in_group_perms(request.user, item)):
			item.user_can_edit = True
		else:
			item.user_can_edit = False

	# Sort editable items to the top of the list
	items = sorted(items, key=lambda x: x.user_can_edit, reverse=True)

	# Pagination
	paginator = Paginator(items, per_page=25)
	paginated_items = paginator.get_page(request.GET.get('p'))

	# Template
	if request.is_ajax():
		template = snippet_type+'/results.html'
	else:
		template = snippet_type+'/type_index.html'

	context = {
		'model_opts': model._meta,
		'items':      paginated_items,
		'can_add_snippet':     request.user.has_perm(get_permission_name('add', model)),
		'can_delete_snippets': request.user.has_perm(get_permission_name('delete', model)),
		'is_searchable': is_searchable,
		'search_form':   search_form,
		'is_searching':  is_searching,
		'query_string':  search_query,
	}

	context = {**context, **add_context}

	return render(request, template, context)

```

**Comment by gasman** [comment]
Hi @juan0tron - have you looked at [ModelAdmin](http://docs.wagtail.io/en/v2.6.3/reference/contrib/modeladmin/index.html) as an alternative?

**Comment by juan0tron** [comment]
@gasman Thanks for the suggestion - ModelAdmin was exactly what I was looking for! Honestly, I'm feeling really dumb right about now. I knew about ModelAdmin months ago and ended up not using it because I couldn't figure it out the first time.

For anyone googling this in the future:

_Make sure you've imported `wagtail.contrib.modeladmin` to your `INSTALLED_APPS`!!_ 

I'm going to close this ticket since it's no longer needed, but I would like to suggest that Wagtail shows a more descriptive error than `TemplateDoesNotExist` if you've registered a modeladmin class without installing the modeladmin app. The only reason I figured this out is because I stumbled upon the solution in [this Stackoverflow thread](https://stackoverflow.com/questions/42355412/templatedoesnotexist-error-for-a-simple-modeladmin-example#42369444) this time around.

This whole time I thought I just wasn't writing my `index.html` file correctly, because I could never get my custom index template to show a list of my snippets. I've been doing this wrong for _nearly a year_ because I failed to realize this. Don't be like me!
